src/Config/ConfigLocator.php

```diff
@@ -400,24 +401,76 @@ public function getSiteAliasPaths($paths, Environment $environment)
      *
      * @return array
      */
-    public function getCommandFilePaths($commandPaths)
+    public function getCommandFilePaths($commandPaths, $root)
     {
-        // Start with the built-in commands.
-        $searchpath = [
+        $builtin = $this->getBuiltinCommandFilePaths();
+        $included = $this->getIncludedCommandFilePaths($commandPaths);
+        $site = $this->getSiteCommandFilePaths(["$root/drush", dirname($root) . '/drush']);
+
+        return array_merge(
+            $builtin,
+            $included,
+            $site
+        );
+    }
+
+    /**
+     * Return all of the built-in commandfile locations
+     */
+    protected function getBuiltinCommandFilePaths()
+    {
+        return [
             dirname(__DIR__),
         ];
+    }
 
+    /**
+     * Return all of the commandfile locations specified via
+     * an 'include' option.
+     */
+    protected function getIncludedCommandFilePaths($commandPaths)
+    {
+        $searchpath = [];
         // Commands specified by 'include' option
         foreach ($commandPaths as $commandPath) {
             if (is_dir($commandPath)) {
                 $searchpath[] = $commandPath;
             }
         }
-
         return $searchpath;
     }
 
     /**
+     * Return all of the commandfile paths in any '$root/drush' or
+     * 'dirname($root)/drush' directory that contains a composer.json
+     * file or a 'Commands' or 'src/Commands' directory.
+     */
```

Comment:
Could we add text that says that CommandFile discovery happens in \Drush\Application::configureAndRegisterCommands. This function is only about discovering paths.